File: cores/Julia/Compositron/src/CDBS/CDBSpectrum.jl
import ..Utils: min_spectrum2d

const Ecal = Tuple{Tuple{Float64, Float64}, Tuple{Float64, Float64}}

mutable struct CDBSpectrum
    spectrum::Matrix{<:Unsigned}
    detpair::String
    ecal::Ecal
    s::Float64
    ds::Float64
    w::Float64
    dw::Float64
    counts::UInt64
    dcounts::Float64
    roi_counts::Float64
    droi_counts::Float64

    function CDBSpectrum(
        spectrum::AbstractMatrix{<:Integer},
        detpair::String,
        ecal::Ecal,
    )
        counts = sum(spectrum)
        dcounts = sqrt(counts)

        c = new()
        c.spectrum = min_spectrum2d(spectrum)
        c.detpair = detpair
        c.ecal = ecal
        c.s = NaN
        c.ds = NaN
        c.w = NaN
        c.dw = NaN
        c.counts = counts
        c.dcounts = dcounts
        c.roi_counts = NaN
        c.droi_counts = NaN

        c
    end
end

File: cores/Python/compositron/src/compositron/cdbs/cdbspectrum.py
import numpy as np
import numpy.typing as npt

from ..core.utils import to_min_type_spectrum


Ecal = tuple[tuple[np.float64, np.float64], tuple[np.float64, np.float64]]

NAN = np.float64(np.nan)


class CDBSpectrum:
    def __init__(
        self,
        spectrum: npt.ArrayLike,
        detpair: str,
        ecal: Ecal,
    ):
        self.spectrum = to_min_type_spectrum(spectrum)
        self.detpair = detpair
        self.ecal = ecal
        self.s = NAN
        self.ds = NAN
        self.w = NAN
        self.ds = NAN
        self.counts = np.sum(self.spectrum, dtype=np.uint64)
        self.dcounts = np.sqrt(self.counts)
        self.roi_counts = NAN
        self.droi_counts = NAN
